insights_content_template_renderer/js_executor.py

```diff
@@ -0,0 +1,166 @@
+                    # Use spawn method to avoid inheriting FastAPI context
+                    ctx = mp.get_context('spawn')
+                    self._process_pool = ctx.Pool(processes=1)
+
+                    log.info("JavaScript worker process initialized successfully")
+
+        return self._process_pool
+
+    def execute(self, js_code, data, timeout=None):
+        """
+        Execute JavaScript code with the given data in a worker process.
+
+        Note: If execution times out, the worker process may still be running.
+        The pool will be terminated and recreated to prevent resource leaks.
+
+        :param js_code: JavaScript code to execute (should be a function)
+        :param data: Data to pass to the JavaScript function
+        :param timeout: Timeout in seconds (default: 5)
+        :return: The result of the JavaScript execution as a string
+        :raises TimeoutError: If execution exceeds timeout
+        :raises RuntimeError: If JavaScript execution fails
+        """
+        if timeout is None:
+            timeout = self._timeout
+
+        try:
+            # Get the process pool (creates it on first call)
+            pool = self.get_pool()
+
+            # Submit task to pool with timeout
+            # Using apply_async allows us to set a timeout without blocking other requests
+            async_result = pool.apply_async(
+                _eval_js_worker_task,
+                args=(js_code, data)
+            )
+
+            # Wait for result with timeout
+            status, result = async_result.get(timeout=timeout)
+
+            if status == 'error':
+                raise RuntimeError(f"JavaScript execution failed: {result}")
+
+            return result
+
+        except mp.TimeoutError:
+            log.error(f"JavaScript execution timed out after {timeout}s, recreating pool")
+
+            # Terminate the pool to kill any hung workers
+            # Note: pythonmonkey doesn't support execution timeouts at the eval() level,
+            # so a hung worker will continue running until terminated
+            if self._process_pool is not None:
+                self._process_pool.terminate()
+                self._process_pool = None
+
+            raise TimeoutError(f"JavaScript execution timed out after {timeout}s")
+
+    def shutdown(self):
+        """
+        Shutdown the process pool gracefully.
+        Should be called during application shutdown.
+        """
+        if self._process_pool is not None:
+            log.info("Shutting down JavaScript worker pool")
+            self._process_pool.close()
+            self._process_pool.join()
+            self._process_pool = None
+            log.info("JavaScript worker pool shutdown complete")
+
+
+# Global singleton instance
+_js_executor = None
+_executor_lock = mp.Lock()
+
+
+def get_js_executor():
+    """
+    Get the global JsExecutor singleton instance.
+
+    :return: The global JsExecutor instance
+    """
+    global _js_executor
```

Comment:
Instead of using global variables, it is better to use `functools.cache`. Return the JsExecutor normally. As it will be cached and only one instance will exist, so no need for _executor_lock (which is weird, it doesn't make much sense here)